runtime/gc_base/ObjectAccessBarrier.cpp

```diff
@@ -1527,7 +1527,31 @@ MM_ObjectAccessBarrier::copyObjectFields(J9VMThread *vmThread, J9Class *objectCl
 			} else {
 				UDATA srcAddress = (UDATA)srcObject + srcOffset + offset;
 				UDATA destAddress = (UDATA)destObject + destOffset + offset;
-				if (sizeof(uint32_t) == referenceSize) {
+				bool copy64Bits = false;
+				UDATA descriptionBitsNext = descriptionBits;
+				const UDATA *descriptionPtrNext = descriptionPtr;
+				UDATA descriptionIndexNext = descriptionIndex;
+				if (isValueType
+					&& (sizeof(uint32_t) == referenceSize)
+					&& (0 == (srcAddress & 7))
+					&& (offset + referenceSize < limit)
+				) {
+					descriptionBitsNext >>= 1;
+					if (descriptionIndexNext-- == 0) {
+						descriptionBitsNext = *(descriptionPtrNext++);
+						descriptionIndexNext = J9_OBJECT_DESCRIPTION_SIZE - 1;
+					}
+					if (0 == (descriptionBitsNext & 1)) {
+						copy64Bits = true;
+					}
+				}
+				if (copy64Bits) {
+					*(uint64_t *)destAddress = *(uint64_t *)srcAddress;
+					descriptionBits = descriptionBitsNext;
+					descriptionPtr = descriptionPtrNext;
+					descriptionIndex = descriptionIndexNext;
+					offset += referenceSize;
```

Comment:
Is this correct? If you copy 64 bits, should you not advance the offset by 8 bytes? I think this code path can be taken when referenceSize == 4.

Reply:
At the end of the while loop (line 1565), the offset is advanced. So for the copy64bit case when referenceSize == 4, offset is advanced 4 bytes here at line 1553 and another 4 byes (taking the normal code path) at line 1565.

Reply:
Please add a comment to this effect.

Reply:
> Please add a comment to this effect.

Done.